### PR TITLE
[Core] Phase out Print packets and added Countdown type to print json

### PR DESCRIPTION
**Pull request by Jarno458**

### CODE REVIEW

File: MultiServer.py [hunk 1]
```diff
@@ -686,22 +686,34 @@ async def on_client_left(ctx: Context, client: Client):
...
-        ctx.broadcast_all([{"cmd": "PrintJSON", "type": "Countdown", "countdown": 0, 
-            "data": [{ "text": "[Server]: GO" }]}])
+        broadcast_countdown(ctx, 0, f"[Server]: GO")
         ctx.countdown_timer = 0
 
 
+def broadcast_countdown(ctx: Context, timer: int, message: str):
+    compatabilityThreshold = Version(0, 3, 5)
+
+    oldClients, newClients = [], []
```
Comment (black-sliver) [inline MultiServer.py:705]: we use snake_case everywhere else in MultiServer.py

**Review by black-sliver — CHANGES_REQUESTED** [review]
just so we don't forget: 0.3.5 is not set in stone, but we should try to make that.

apcpp seems to ignore PrintJSON with unknown type, so no output
no clue about js and java libs
commonclient works according to berserker
apclientpp should work as long as you use the render_json function to stringify it

also normally we have a comment like "remove backwards compatibility around 0.3.7"

**Review by black-sliver — COMMENTED** [review]
Looking good, we can't merge it until we know 0.3.5 is gonna work for sure though

File: docs/network protocol.md [hunk 2]
```diff
@@ -164,10 +164,19 @@ Sent to clients purely to display a message to the player. This packet differs f
 | Name | Type | Notes |
 | ---- | ---- | ----- |
 | data | list\[[JSONMessagePart](#JSONMessagePart)\] | Type of this part of the message. |
-| type | str | May be present to indicate the nature of this message. Known types are Hint and ItemSend. |
+| type | str | May be present to indicate the [PrintJsonType](#PrintJsonType) of this message. |
 | receiving | int | Is present if type is Hint or ItemSend and marks the destination player's ID. |
 | item | [NetworkItem](#NetworkItem) | Is present if type is Hint or ItemSend and marks the source player id, location id, item id and item flags. |
 | found | bool | Is present if type is Hint, denotes whether the location hinted for was checked. |
+| countdown | int | Is present if type is Countdown, denotes the amount of seconds remaining on the countdown |
```
Comment (Ijwu) [inline docs/network protocol.md:171]:
```suggestion
| countdown | int | Is present if type is `Countdown`, denotes the amount of seconds remaining on the countdown. |
```

File: docs/network protocol.md [hunk 3]
```diff
@@ -164,10 +164,19 @@ Sent to clients purely to display a message to the player. This packet differs f
...
+
+##### PrintJsonType
+PrintJsonType indicates the type of [PrintJson](#PrintJson) packet, different types can be handled differently by the client and can also contain different additional arguments. When receiving an unknown type the data's list\[[JSONMessagePart](#JSONMessagePart)\] should still be printed as normally. Currently known types:
+| Type | Notes |
+| ---- | ----- |
+| ItemSend | The message is in response to a player receiving an item |
```
Comment (Ijwu) [inline docs/network protocol.md:177]:
```suggestion
| ItemSend | The message is in response to a player receiving an item. |
```

File: docs/network protocol.md [hunk 4]
```diff
@@ -164,10 +164,19 @@ Sent to clients purely to display a message to the player. This packet differs f
...
 | data | list\[[JSONMessagePart](#JSONMessagePart)\] | Type of this part of the message. |
-| type | str | May be present to indicate the nature of this message. Known types are Hint and ItemSend. |
+| type | str | May be present to indicate the [PrintJsonType](#PrintJsonType) of this message. |
 | receiving | int | Is present if type is Hint or ItemSend and marks the destination player's ID. |
 | item | [NetworkItem](#NetworkItem) | Is present if type is Hint or ItemSend and marks the source player id, location id, item id and item flags. |
 | found | bool | Is present if type is Hint, denotes whether the location hinted for was checked. |
+| countdown | int | Is present if type is Countdown, denotes the amount of seconds remaining on the countdown |
+
+##### PrintJsonType
+PrintJsonType indicates the type of [PrintJson](#PrintJson) packet, different types can be handled differently by the client and can also contain different additional arguments. When receiving an unknown type the data's list\[[JSONMessagePart](#JSONMessagePart)\] should still be printed as normally. Currently known types:
```
Comment (Ijwu) [inline docs/network protocol.md:174]:
```suggestion
PrintJsonType indicates the type of [PrintJson](#PrintJson) packet, different types can be handled differently by the client and can also contain additional arguments. When receiving an unknown type the data's list\[[JSONMessagePart](#JSONMessagePart)\] should still be printed as normal. 

Currently defined types are:
```

File: docs/network protocol.md [hunk 5]
```diff
@@ -164,10 +164,19 @@ Sent to clients purely to display a message to the player. This packet differs f
...
+| Type | Notes |
+| ---- | ----- |
+| ItemSend | The message is in response to a player receiving an item |
+| Hint | The message is in response to a player hinting |
```
Comment (Ijwu) [inline docs/network protocol.md:178]:
```suggestion
| Hint | The message is in response to a player hinting. |
```

File: docs/network protocol.md [hunk 6]
```diff
@@ -164,10 +164,19 @@ Sent to clients purely to display a message to the player. This packet differs f
...
+| Countdown | The message contains information about the current server Countdown |
```
Comment (Ijwu) [inline docs/network protocol.md:179]:
```suggestion
| Countdown | The message contains information about the current server Countdown. |
```

**Review by Ijwu — COMMENTED** [review]
Looks good, awaiting some form of feedback based on @black-sliver's comments.

**Comment by black-sliver** [comment]
> Looks good, awaiting some form of feedback based on @black-sliver's comments.

I agree with all doc change suggestions, thanks for taking your time.

Not sure how to proceed wrt to client lib changes. I feel like we should try to get past RC first, and then annoy people :)

**Comment by Jarno458** [comment]
Not sure why this was tagged help wanted, as I think that is something you use when you want other devs to help finish it

**Comment by black-sliver** [comment]
> Not sure why this was tagged help wanted, as I think that is something you use when you want other devs to help finish it

because we need someone to test Java and JS and someone to fix apcpp? I can also just @KonoTyran and @PoryGone and @ThePhar I guess

**Comment by PoryGone** [comment]
> > Not sure why this was tagged help wanted, as I think that is something you use when you want other devs to help finish it
> 
> because we need someone to test Java and JS and someone to fix apcpp? I can also just @KonoTyran and @PoryGone and @ThePhar I guess

I'll be open to helping with APCpp, but @N00byKing gets first dibs on changes to that lib.
Might be good to summarize what actually needs doing here though, as there are a ton of threads in this PR, and it's not clear what the final result is.

**Comment by Jarno458** [comment]
So to summerize, 

If the client sends Version 0.3.4 or lower, nothing changes
If the client sends Version 3.3.5 or greater. than during a `/countdown` the client will not be updated with Print packets but instead receive PrintJSON packets

**Comment by KonoTyran** [comment]
The Java lib doesn't actually care. It actually sends most of it off to the game to handle as it has no way of knowing how the game will display the text.
Minecraft doesn't actually look at the type, and sts ignores all pringjsons

**Comment by black-sliver** [comment]
> I'll be open to helping with APCpp, but @N00byKing gets first dibs on changes to that lib.

Sorry, might've been too early for my brain to function. 

Basically PrintJSON should behave the same as Print if the type is unknown (and you can have special handling for countdown if you want to). We'll hopefully have the same change for Say to completely get rid of the old Print.

You can track the overall planned protocol changes in issue #813 and read the [changes to network protocol.md](https://github.com/ArchipelagoMW/Archipelago/pull/812/files#diff-2fdfc3f0c7765b3008f036da659a2b55754707a68a56e27b396fedb902e2cbe9) for this particular part of it.

**Comment by Jarno458** [comment]
As the new title reflects, this was extended to remove all PrintPackets

File: MultiServer.py [hunk 7]
```diff
@@ -685,19 +693,33 @@ async def on_client_left(ctx: Context, client: Client):
...
+        for clients in teams.values():
+            for client in clients:
+                new_clients.append(client) if client.version >= print_command_compatability_threshold \
+                    else old_clients.append(client)
+
+    ctx.broadcast(old_clients, [{"cmd": "Print", "text": text }])
+    ctx.broadcast(new_clients, [{**{"cmd": "PrintJSON", "data": [{ "text": text }]}, **additional_arguments}])
+
+def broadcast_countdown(ctx: Context, timer: int, message: str):
+    broadcast_text_all(ctx, message, { "type": "Countdown", "countdown": timer })
```
Comment (black-sliver) [inline MultiServer.py:722]: could drop the message as argument and instead do `f"[Server]: {n if n else 'Go'}"` instead

Comment (Jarno458) [reply MultiServer.py:722]: Well, there is also the first message
"[Server]: Starting countdown of {timer}s"

Comment (black-sliver) [reply MultiServer.py:722]: you are absolutely right. i'm just blind

**Comment by N00byKing** [comment]
I've rewritten the way APCpp handles messages in a way compatible with this PR (on a seperate branch for now).

While you're at it, I'd really like if the introductory texts (`X playing Y has joined` and `Now that you are connected, ...`) can be given a type as well, just to have a way to filter them out. They're quite long and contain special characters that sm64 doesnt like.

**Comment by Jarno458** [comment]
could easily add an type but not sure on how to call them

**Comment by black-sliver** [comment]
@N00byKing make sure to have the library send 0.3.5 if it's supported and <0.3.5 if not (i.e. update the default if you have one).

@Jarno458 could be "room_update" for joined/left and "help" for "Now that you are connected..."

**Comment by PoryGone** [comment]
I've seen that @N00byKing pushed updates to APCpp in regards to this PR. Is it necessary for this update that SA2 push a release with that updated client?

**Comment by N00byKing** [comment]
Not with this PR; it keeps the (soon to be removed) old path.
When that one gets removed the old version will continue to work in limited capacity; notably some messages will be missing. It wont break spectacularly though

**Comment by black-sliver** [comment]
> Is it necessary for this update that SA2 push a release with that updated client?

Apcpp users can't change the version they send unless they also update apcpp - otherwise they will miss out on a lot of messages.
(The same or a similar thing may be true for clients that were not tested, it should be noticed during testing when updating the client though.)

**Review by black-sliver — APPROVED** [review]
Works on my end and code looks good

(Note the compatibility warnings above; clients can't send 0.3.5 version without properly supporting PrintJSON)

**Pull request merged**